src/shogun/mathematics/linalg/internal/implementation/Variance.h

```diff
@@ -0,0 +1,319 @@
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef SRC_SHOGUN_MATHEMATICS_LINALG_INTERNAL_IMPLEMENTATION_VARIANCE_H_
+#define SRC_SHOGUN_MATHEMATICS_LINALG_INTERNAL_IMPLEMENTATION_VARIANCE_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/mathematics/linalg/internal/implementation/MeanEigen3.h>
+#include <shogun/mathematics/eigen3.h>
+
+#include <iostream>
+using namespace std;
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+namespace implementation
+{
+
+/**
+ * @brief Generic class variance which provides a static compute method. This class
+ * can work with generic matricies.
+ */
+template <enum Backend, class Matrix>
+struct variance
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Calculates unbiased empirical variance estimate given the entries from a matrix. Given a matrix
+	 * \f$x\f$ with entries \f$\{x_{11}, ..., x_{mn}\}\f$, this is
+	 * \f$\frac{1}{m*n-1}\sum_{i=1}^m\sum_{j=1}^n (x_{ij}-\bar{x})^2\f$ where
+	 * \f$\bar x=\frac{1}{mn}\sum_{i=1}^m\sum_{j=1}^n x_{ij}\f$
+	 *
+	 * @param x matrix of values
+	 * @return variance of given values
+	 */
+	static T compute(Matrix x);
+};
+
+/**
+ * @brief Specialization of element-wise variance which works with SGMatrix
+ * and uses Eigen3 as backend for computing variance.
+ */
+template <class Matrix>
+struct variance<Backend::EIGEN3, Matrix>
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Eigen matrix type */
+	typedef Eigen::Matrix<T,Eigen::Dynamic, Eigen::Dynamic> MatrixXt;
+
+	/** Calculates unbiased empirical variance estimate given the entries from a matrix. Given a matrix
+	 * \f$x\f$ with entries \f$\{x_{11}, ..., x_{mn}\}\f$, this is
+	 * \f$\frac{1}{m*n-1}\sum_{i=1}^m\sum_{j=1}^n (x_{ij}-\bar{x})^2\f$ where
+	 * \f$\bar x=\frac{1}{mn}\sum_{i=1}^m\sum_{j=1}^n x_{ij}\f$
+	 *
+	 * @param x matrix of values
+	 * @return variance of given values
+	 */
+	static T compute(SGMatrix<T> x)
+	{
+		REQUIRE(x.num_rows > 0, "Please ensure that m has more than 0 rows.\n")
+		REQUIRE(x.num_cols > 0, "Please ensure that m has more than %d columns.\n")
+
+		SGMatrix<T> squaredResult(x.num_rows, x.num_cols);
+		Eigen::Map<MatrixXt> eigX = x;
+		Eigen::Map<MatrixXt> eigSquaredResult = squaredResult;
+
+		T meanVal = mean<Backend::EIGEN3, SGMatrix<T>>::compute(x);		
+		eigSquaredResult.fill(meanVal);
+		eigSquaredResult = eigX - eigSquaredResult;
```

Comment:
This line could be problematic. Did you make sure that it works?

Also, since this squared result thing is not being returned by the API, it can just be a Eigen3 matrix that you use internally for this method. As long as the API doesn't take or return a non SG type, we're good. See what I mean?